src/sql/azdata.proposed.d.ts

```diff
@@ -2073,6 +2073,436 @@ declare module 'azdata' {
+	 */
+	export interface ChartData<TChartType extends ChartType> {
+		/**
+		 * Never needs to be set or used.  Only present for the TypeScript compiler to recognize the pairing between same-chart Data and Options types.
+		 */
+		// DevNote:
+		// This works because it gets compiled to (e.g.) `'bar' | undefined, forcing it to be associated with BarChartOptions
+		// and preventing it from being paired with PieChartOptions.
+		type?: TChartType;
+	}
+
+	//#region Chart general data types
+
+	export interface ChartDataEntryBase {
+		/**
+		 * For Pie, Doughnut, Polar Area charts, it the label associated with the data value
+		 * For Bar, Horizontal Bar, Line, Scatterplot, Bubble, and Rradial, it is the label name for dataset
+		 */
+		dataLabel: string;
+		/**
+		 * Background color for chart elements.  May be a name ('red'), hex ('#FFFFFF[77]), or RGB ('rgb(255, 255, 255[, 0.5])).
+		 * Bracketed portions are optional for alpha/transparency.
+		 */
+		backgroundColor: string;
+		/**
+		 * Border color for chart elements.  May be a name ('red'), hex ('#FFFFFF[77]), or RGB ('rgb(255, 255, 255[, 0.5])).
+		 * Bracketed portions are optional for alpha/transparency.
+		 */
+		borderColor?: string;
+	}
+
+	export interface ChartDataEntry extends ChartDataEntryBase {
+		/**
+		 * Value of one-dimensional data point
+		 */
+		value: Chart1DPoint | number;
+	}
+
+	export interface ChartDataSet<TVal extends Chart1DPoint | number> extends ChartDataEntryBase {
+		data: TVal[];
+	}
+
+	/**
+	 * One-dimensional data point
+	 */
+	export interface Chart1DPoint {
+		/**
+		 * Value for a one-dimensional data point, or the x-coordinate for a multi-dimensional data point
+		 */
+		x: number;
+	}
+
+	/**
+	 * Two-dimensional data point
+	 */
+	export interface Chart2DPoint extends Chart1DPoint {
+		/**
+		 * Y-coordiate for a multi-dimensional data point
+		 */
+		y: number;
+	}
+
+	/**
+	 * Three-dimensional data point
+	 */
+	export interface Chart3DPoint extends Chart2DPoint {
+		/**
+		 * Radius for a bubble data point, in pixels
+		 */
+		r: number;
+	}
+
+	//#endregion
+
+	//#region Chart general option types
+
+	/**
+	 * Base options for a chart
+	 */
+	export interface ChartOptions<TChartType extends ChartType> {
+		/**
+		 * Never needs to be set or used.  Only present for the TypeScript compiler to recognize the pairing between same-chart Data and Options types.
+		 */
+		// DevNote:
+		// This works because it gets compiled to (e.g.) `'bar' | undefined, forcing it to be associated with BarChartData
+		// and preventing it from being paired with PieChartData.
+		type?: TChartType;
+
+		/**
+		 * Title of the chart.  Set to `undefined` to not display the title.
+		 */
+		chartTitle?: string;
+
+		/**
+		 * Whether to display the legend.  Defaults to true.
+		 */
+		legendVisible?: boolean;
+
+		/**
+		 * Free-form options that are mixed into the display options for the underlying charting library to support all missing options functionality.
```

Comment:
Is this really necessary right now? The problem is that if we switch libraries, then suddenly we have to worry about breaking any extensions that use this. Having strongly typed properties that we control is better - we can always add more if needed. That's the whole point of making this API generic - the caller isn't supposed to know (or care) how the chart is rendered directly as long as the behavior is consistent.

Reply:
I think it's a good addition, and is one that I made after already considering the ramifications of switching libraries.  My thinking is that the explicitly-defined options are those that we're guaranteeing support for in the event of changing libraries, and that - if they accept the caveat in the docstring - they can delve into the deeper, undocumented-in-ADS options for full customization.

Anything getting set via this is 1) a non-mission-critical option that 2) won't impact the actual usability of the chart, and 3) the developer acknowledges may break if we change libraries.  I think that with appropriate messaging in the docstring, that's okay, but maybe I can word the warning more strongly.  Or maybe we can mark this as `@deprecated` from the beginning, then elaborate in the rest of the comment that we really just mean "subject to change without warning".  But I _do_ think this is a good option to keep.

Reply:
Do you have a scenario that actually warrants that? Until we do I would prefer we leave it out. We can always add it in later if it seems necessary.

The API is purposely supposed to be abstracted. Anything that detracts from that is going against the goal of having it be generic in the first place and will make any backend updates we do harder without risking causing unintended changes in extensions using this API.

Reply:
Discussed briefly offline.  I've removed the ability to set free-form options, and will bring it up at a future tech sync to field options.